pageserver/src/tenant/secondary/scheduler.rs

```diff
@@ -0,0 +1,396 @@
+use async_trait;
+use futures::{Future, FutureExt};
+use std::{
+    collections::HashMap,
+    marker::PhantomData,
+    pin::Pin,
+    time::{Duration, Instant},
+};
+
+use pageserver_api::shard::TenantShardId;
+use tokio::task::JoinSet;
+use tokio_util::sync::CancellationToken;
+use utils::completion::Barrier;
+
+use super::{CommandRequest, CommandResponse};
+
+/// Scheduling interval is the time between calls to JobGenerator::schedule.
+/// When we schedule jobs, the job generator may provide a hint of its preferred
+/// interval, which we will respect within these intervals.
+const MAX_SCHEDULING_INTERVAL: Duration = Duration::from_secs(10);
+const MIN_SCHEDULING_INTERVAL: Duration = Duration::from_secs(1);
+
+/// Scheduling helper for background work across many tenants.
+///
+/// Systems that need to run background work across many tenants may use this type
+/// to schedule jobs within a concurrency limit, along with their own [`JobGenerator`]
+/// implementation to provide the work to execute.  This is a simple scheduler that just
+/// polls the generator for outstanding work, replacing its queue of pending work with
+/// what the generator yields on each call: the job generator can change its mind about
+/// the order of jobs between calls.  The job generator is notified when jobs complete,
+/// and additionally may expose a command hook to generate jobs on-demand (e.g. to implement
+/// admin APIs).
+///
+/// For an example see [`crate::tenant::secondary::heatmap_uploader`]
+///
+/// G: A JobGenerator that this scheduler will poll to find pending jobs
+/// PJ: 'Pending Job': type for job descriptors that are ready to run
+/// RJ: 'Running Job' type' for jobs that have been spawned
+/// C : 'Completion' type that spawned jobs will send when they finish
+/// CMD: 'Command' type that the job generator will accept to create jobs on-demand
+pub(super) struct TenantBackgroundJobs<G, PJ, RJ, C, CMD>
+where
+    C: TenantScoped + Send,
+    PJ: TenantScoped,
+    RJ: HasBarrier,
+    G: JobGenerator<PJ, RJ, C, CMD>,
```

Comment:
I am skeptical if we need this kind of infra, but I will try to find out why this was added and how that could be simplified.

Reply:
This is a generalization of the heatmap uploader's  job code, to avoid implementing anything separate for the secondary downloader.

The reasons for using a scheduler like this rather than the traditional task-per-tenant-plus-semaphore approach are:
- Prioritized scheduling: rather than tasks fighting for a semaphore and running on a first-come-first-served, basis, on each scheduling interval the JobGenerator  can order them as it wishes: an upload/download with a deadline very soon will take priority over one that just happened to start grabbing the semaphore sooner.  This is quite simple at the  moment but the overall structure enables arbitry prioritization in future, such as prioritizing more busy/less busy tenants.  Imagine if we were using this for compaction in future (not saying we have to, but for example): we could prioritize the tenants with the longest backlog of deltas.
- Avoid a large population of idle tasks waiting for their next periodic work (tasks are relatively cheap, but this will be helpful as/when we have more instrumentation around tokio tasks and want to understand the tasks we've got)

In terms of complexity, I think it's quite modest: scheduler.rs is <400 lines of code including comments, and the JobGenerator trait only has four methods.  As scheduling code for background jobs goes, this is a very lightweight one.